### DISCUSSION TITLE
Using google/highway for vectorization instead of depending on the compiler autovectorization

### DISCUSSION THREAD

**Discussion by krishvishal** [mlpack/mlpack#3698] [Ideas]
I've seen few loops marked as opportunities for SIMD autovectorization in the repo. As the title says, what do you think about using [google/highway](https://github.com/google/highway/) for this purpose. It has dynamic dispatch to auto generate code for a specific CPU architecture.

highway also uses C++14.

**Comment by rcurtin** [comment]
Hey there @krishvishal, it's a cool project, thanks for linking to it.  But personally I really would prefer to see us avoid adding new dependencies---in the past few years, we've gone through a lot of effort to remove dependencies so that mlpack is easier to configure and use on low-resource devices.  From a quick look it would be a huge amount of effort to refactor to use it well, and it's unclear to me how well (or not) highway would interact with Armadillo or Bandicoot for linear algebra.

**Comment by krishvishal** [comment]
Thank you for replying.

What performance opportunities do you see in this project? I'm keen to work on them.
